pennylane/devices/task_qubit.py

```diff
@@ -0,0 +1,372 @@
+    as mentioned at https://stackoverflow.com/questions/28237955/same-name-for-classmethod-and-instancemethod/28238047#28238047
+    """
+
+    def __init__(self, fclass, finstance=None, doc=None):
+        self.fclass = fclass
+        self.finstance = finstance
+        self.__doc__ = doc or fclass.__doc__
+        self.__isabstractmethod__ = bool(getattr(fclass, "__isabstractmethod__", False))
+
+    def classmethod(self, fclass):
+        """
+        Defines the decorated method as a class-method.
+        """
+        return type(self)(fclass, self.finstance, None)
+
+    def instancemethod(self, finstance):
+        """
+        Defines the decorated method as an instance-method.
+        """
+        return type(self)(self.fclass, finstance, self.__doc__)
+
+    def __get__(self, instance, cls):
+        if instance is None or self.finstance is None:
+            return self.fclass.__get__(cls, None)
+        return self.finstance.__get__(instance, cls)
+
+
+class TaskQubit(DefaultQubit):
+    """Proxy simulator plugin written using Dask.Distributed as a task-distribution scheduling backend.
+
+    **Short name:** ``task.qubit``
+
+    This device provides a pure-state qubit simulator wrapping both ``"default.qubit"`` and ``"lightning.qubit"``,
+    and written to allow batched offloading to a Dask scheduler. The ``task.qubit`` device works with the Autograd,
+    TensorFlow, PyTorch and (non-JIT) JAX interfaces. Currently, support exists for both parameter-shift and
+    backpropagation differentation methods; adjoint support is not currently enabled.
+
+    To use this device, you will need to install dask and dask.distributed:
+
+    .. code-block:: console
+
+        pip install dask distributed
+
+    **Example**
+
+    >>> import pennylane as qml
+    >>> import pennylane.numpy as np
+    >>> import tensorflow as tf
+    >>> import dask.distributed as dist
+
+    >>> if __name__ == '__main__': # Required for LocalCluster
+    ...     cluster = dist.LocalCluster(n_workers=4, threads_per_worker=1)
+    ...     client = dist.Client(cluster)
+    ...     backend = "default.qubit"
+    ...     dev = qml.device("task.qubit", wires=6, backend=backend)
+    ...     @qml.qnode(dev, cache=False, interface="tf", diff_method="parameter-shift") # caching must be disabled due to proxy interface
+    ...     def circuit(x):
+    ...         qml.RX(x[0], wires=0)
+    ...         qml.RY(x[1], wires=0)
+    ...         qml.RZ(x[2], wires=0)
+    ...         qml.RZ(x[0], wires=1)
+    ...         qml.RX(x[1], wires=1)
+    ...         qml.RY(x[2], wires=1)
+    ...         return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+    ...     weights = tf.Variable(np.random.rand(3))
+    ...     def f_submit(weights):
+    ...         with tf.GradientTape() as tape:
+    ...             # Use the circuit to calculate the loss value
+    ...             loss = tf.abs(circuit(weights)-0.5)**2
+    ...         return tape.gradient(loss, weights)
```

Comment:
This is one of the sharp edges --- the problem here (I think) is that the weights exist on the host, but the computation happens on the worker. Since the memory is not explicitly copied, TF does not see the data as being the same, and we get the `tf.UnconnectedGradients` return type. I have yet to find a good solution to this.

The only way around this right now, is to package everything into a callable, and submit that. Avoiding the REPL seems to help here.
```
import pennylane as qml
import pennylane.numpy as np
import tensorflow as tf
import dask

if __name__ == '__main__':

    cluster = dask.distributed.LocalCluster(n_workers=4, threads_per_worker=1)
    client =  dask.distributed.Client(cluster)

    def my_func(weights_host):
        qpu = qml.device(
            "task.qubit",
            wires=3,
            backend="lightning.qubit",
            future=False
        )

        H = -1.0 * qml.PauliX(0) @ qml.PauliZ(1) + 0.5 * qml.PauliY(2)

        @qml.qnode(qpu, cache=False, interface="tf")
        def circuit1(params):
            qml.StronglyEntanglingLayers(params, wires=[0, 1, 2])
            return qml.expval(H)

        @qml.qnode(qpu, cache=False, interface="tf")
        def circuit2(params):
            qml.StronglyEntanglingLayers(params, wires=[0, 1, 2])
            return qml.probs(wires=0)

        weights = tf.Variable(weights_host)

        def cost(weights):
            return circuit1(weights) * tf.reduce_sum(tf.sin(circuit2(weights)))

        with tf.GradientTape() as tape:
            loss = cost(weights)

        grad = tape.gradient(loss, weights)
        return grad

    f = client.submit(my_func, np.random.random([2, 3, 3]))
    print(f.result())
```

Reply:
I suppose you could also submit the optimization loop, but that probably doesn't distribute so well 🤔 This is super strange, do we know for sure if Dask is supposed to be able to support distributed TF graphs?

Reply:
I think this is expected. There is no internal support to automatically take a TF graph and handle distribution --- Dask doesn't operate as a distribution framework, more akin to a task-based scheduler. We can explicitly tell it what to run, and tell it which data depend on another, but it doesn't have support to reason what it can and cannot do.

If we wished to have the full graph executed, we would need to define serialization for it, define a dependency graph within Dask, and a wait of enqueing the ops. This is the same as the other frameworks. Dask's potential here is more for the ability to execute asynchronously and efficiently scheduling what we ask (in a similar manner to a smarter `concurrent.futures`).